File: hackerrank/problem solving/algorithms/2 - implementation/50 - chocolate feast/chocolate_feast.c
// https://www.hackerrank.com/challenges/chocolate-feast/problem?isFullScreen=true

#include <stdio.h>

int how_many_chocolates_can_be_eaten(int money, int cost, int wrappers_needed);

int main()
{
    unsigned n;
    scanf("%u", &n);

    while (n--) {
        int money, cost, wrappers_needed;
        scanf("%d %d %d", &money, &cost, &wrappers_needed);
        printf("%d\n", how_many_chocolates_can_be_eaten(money, cost, wrappers_needed));
    }

    return 0;
}

int how_many_chocolates_can_be_eaten(int money, int cost, int wrappers_needed)
{
    int chocolates = money / cost;

    for (int wrappers = chocolates; wrappers >= wrappers_needed;) {
        int free_chocolates = wrappers / wrappers_needed;
        wrappers = wrappers % wrappers_needed + free_chocolates;
        chocolates += free_chocolates;
    }

    return chocolates;
}

File: hackerrank/problem solving/algorithms/2 - implementation/50 - chocolate feast/chocolate_feast.cs
// https://www.hackerrank.com/challenges/chocolate-feast/problem?isFullScreen=true

int n = int.Parse(Console.ReadLine()!);
for (int i = 0; i < n; i++)
{
    var arr = ReadNumbers();
    int money = arr[0];
    int cost = arr[1];
    int wrappersNeeded = arr[2];
    Console.WriteLine(HowManyChocolatesCanBeEaten(money, cost, wrappersNeeded));
}

int[] ReadNumbers() => Console.ReadLine()!.Split().Select(int.Parse).ToArray();

// T: O(log(money / cost))
// S: O(1) extra space
int HowManyChocolatesCanBeEaten(int money, int cost, int wrappersNeeded)
{
    int chocolates = money / cost;

    for (int wrappers = chocolates; wrappers >= wrappersNeeded;)
    {
        int freeChocolates = wrappers / wrappersNeeded;
        wrappers = wrappers % wrappersNeeded + freeChocolates;
        chocolates += freeChocolates;
    }

    return chocolates;
}
